module/tty0tty.c

```diff
@@ -91,9 +91,52 @@ struct tty0tty_serial {
 static struct tty0tty_serial *tty0tty_table[TTY0TTY_MINORS];	/* initially all NULL */
 
 
+static struct tty0tty_serial *get_shadow_tty(int index)
+{
+	struct tty0tty_serial *shadow = NULL;
+	int shadow_idx = 1 - (index % 2);
```

Comment:
for index pairs 0 and 1 it works, but for pairs 2 and 3, 4 and 5 I don't think so.

Reply:
int shadow_idx = index + ((index+1) % 2) - (index % 2) ; its solve. Can you better code this?

Reply:
Sorry.  Good catch.  I've fixed the shadow_idx line:
`int shadow_idx = index ^ 1;`

Much cleaner.

I've tried it on kernel 4.15.0-106 (Ubuntu 18.04 LTS) and 4.4.0-184 (Ubuntu 16.04 LTS) (both x86_64).  Neither worked without these changes.  I'm not sure how it _could_ work since the work queue was not initialised, the counters were not incremented and there was nothing to wake up the work queue.  Certainly anything using the tiocmiwait ioctl is unlikely to work unless some new magic has been done in v5 kernels (although I'm still not sure what that might be).  Maybe I've missed something?

Reply:
Can you send me the code you are using to test? 
When I said it works it was about the connection between RTS, CTS signals, etc ...
I haven't tested it with ioctl tiocmiwait, the implementation you made is apparently correct.
 I am not an expert in developing linux drivers, after 8 years you are the first person to contribute with improvements.

Reply:
I just added a simple test program to https://github.com/rhulme/tty0tty/tree/feature/signal_test.  See if it works for you.

It's nothing pretty but it should be enough to test the feature

Reply:
All Ok. Thanks for the help.